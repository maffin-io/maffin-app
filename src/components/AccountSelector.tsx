import React from 'react';
import { useHotkeys } from 'react-hotkeys-hook';
import Select, {
  SingleValue,
  components,
} from 'react-select';
import { BiSearch } from 'react-icons/bi';
import useSWRImmutable from 'swr/immutable';

import { Account } from '@/book/entities';
import { getAccountsWithPath } from '@/book/queries';

export type AccountSelectorProps = {
  id?: string,
  placeholder?: string,
  onChange?: Function,
  showRoot?: boolean,
  isClearable?: boolean,
  ignoreAccounts?: string[],
  defaultValue?: Account,
  className?: string,
};

export default function AccountSelector(
  {
    placeholder,
    onChange,
    ignoreAccounts,
    id = 'accountSelector',
    showRoot = false,
    isClearable = true,
    defaultValue,
    className = '',
  }: AccountSelectorProps,
): JSX.Element {
  const ref = React.useRef<HTMLDivElement>();
  useHotkeys('meta+k', () => {
    ref.current?.focus();
  });

  let { data: accounts } = useSWRImmutable(
    '/api/accounts',
    getAccountsWithPath,
  );

  accounts = accounts || [];
  accounts = accounts.filter(account => !(ignoreAccounts || []).includes(account.type));
  if (!showRoot) {
    accounts = accounts.filter(account => account.type !== 'ROOT');
  }

  return (
    <Select
      // @ts-ignore
      ref={ref}
      id={id}
      name={id}
      aria-label={id}
      options={accounts}
      placeholder={placeholder || 'Choose account'}
      onChange={(newValue: SingleValue<Account> | null) => {
        if (onChange) {
          onChange(newValue);
        }
      }}
      isClearable={isClearable}
      defaultValue={defaultValue}
      backspaceRemovesValue
      onKeyDown={(e: React.KeyboardEvent) => {
        if (e.key === 'Escape') {
          ref.current?.blur();
        }
      }}
      getOptionLabel={(option) => option.path}
      getOptionValue={(option) => option.path}
      openMenuOnFocus
      components={{ Control }}
      className={className}
      styles={{
        option: () => ({}),
      }}
      classNames={{
        // We set h-9 so it's the same as other inputs
        control: () => '!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none',
        indicatorSeparator: () => 'hidden',
        menu: () => 'text-sm !bg-gunmetal-800',
        option: (state) => {
          const styles = 'w-full cursor-default hover:bg-gunmetal-700 px-4 py-2';
          if (state.isFocused) {
            // For some reason putting focus:bg-gunmetal-700 doesnt work
            return `bg-gunmetal-700 ${styles}`;
          }
          return styles;
        },
        singleValue: () => '!text-inherit',
        input: () => '!text-inherit',
      }}
    />
  );
}

function Control(
  {
    children,
    ...props
  }: any,
): JSX.Element {
  return (
    <components.Control {...props}>
      <BiSearch className="absolute text-md left-[10px]" />
      {children}
    </components.Control>
  );
}
