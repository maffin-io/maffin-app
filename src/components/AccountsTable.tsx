import React from 'react';
import { ColumnDef } from '@tanstack/react-table';
import Link from 'next/link';
import { BiCircle, BiSolidRightArrow, BiSolidDownArrow } from 'react-icons/bi';
import classNames from 'classnames';

import Table from '@/components/Table';
import type { AccountsTree } from '@/types/accounts';

export type AccountsTableProps = {
  tree: AccountsTree,
};

export default function AccountsTable(
  {
    tree,
  }: AccountsTableProps,
): JSX.Element {
  return (
    <Table<AccountsTree>
      columns={columns}
      data={tree.children}
      initialSort={{ id: 'total', desc: true }}
      showHeader={false}
      showPagination={false}
      tdClassName="p-2"
      getSubRows={row => row.children}
    />
  );
}

const columns: ColumnDef<AccountsTree>[] = [
  {
    accessorKey: 'name',
    id: 'name',
    enableSorting: false,
    header: '',
    cell: ({ row }) => (
      <div
        className={`flex pl-${row.depth * 4}`}
      >
        {row.getCanExpand() ? (
          <button
            type="button"
            onClick={() => row.toggleExpanded()}
          >
            {(
              row.getIsExpanded()
                ? <BiSolidDownArrow className="mr-1 text-xs opacity-20" />
                : <BiSolidRightArrow className="mr-1 text-xs opacity-20" />
            )}
          </button>
        ) : (
          <button
            type="button"
            className="cursor-default"
          >
            <BiCircle className="mr-1 text-xs opacity-20" />
          </button>
        )}
        <Link
          className={classNames('badge hover:text-slate-300', {
            'bg-green-500/20 text-green-300': row.original.account.type === 'INCOME',
            'bg-red-500/20 text-red-300': row.original.account.type === 'EXPENSE',
            'bg-cyan-500/20 text-cyan-300': ['ASSET', 'BANK'].includes(row.original.account.type),
            'bg-orange-500/20 text-orange-300': row.original.account.type === 'LIABILITY',
            'bg-violet-500/20 text-violet-300': ['STOCK', 'MUTUAL'].includes(row.original.account.type),
          })}
          href={`/dashboard/accounts/${row.original.account.guid}`}
        >
          {row.original.account.name}
        </Link>
      </div>
    ),
  },
  {
    accessorFn: (row: AccountsTree) => row.total.toNumber(),
    id: 'total',
    header: '',
    cell: ({ row }) => (
      <span>
        {row.original.total.format()}
      </span>
    ),
  },
];
