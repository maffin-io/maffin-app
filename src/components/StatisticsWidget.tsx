import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  className?: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  className = '',
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className={`${className} bg-gunmetal-700 rounded-sm p-6`}>
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
