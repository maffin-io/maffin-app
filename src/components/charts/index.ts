export { default as Bar } from './Bar';
export { default as Line } from './Line';
export { default as Tree } from './Tree';
export { default as Pie } from './Pie';

export { default as NetWorthHistogram } from './NetWorthHistogram';
