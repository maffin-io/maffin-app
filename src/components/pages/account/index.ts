export { default as TotalLineChart } from './TotalLineChart';
export { default as SplitsHistogram } from './SplitsHistogram';
export { default as TransactionsTable } from './TransactionsTable';
export { default as Header } from './Header';
export { default as AccountInfo } from './AccountInfo';
export { default as InvestmentInfo } from './InvestmentInfo';
export { default as AssetInfo } from './AssetInfo';
