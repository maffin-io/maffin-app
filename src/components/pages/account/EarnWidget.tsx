import React from 'react';
import { DateTime, Interval } from 'luxon';

import { useMainCurrency, useCashFlow } from '@/hooks/api';
import type { Account } from '@/book/entities';
import StatisticsWidget from '@/components/StatisticsWidget';
import { moneyToString } from '@/helpers/number';

export type EarnWidgetProps = {
  account: Account,
};

export default function EarnWidget({
  account,
}: EarnWidgetProps): JSX.Element {
  const { data: currency } = useMainCurrency();
  const { data: cashflow0 } = useCashFlow(account.guid);
  const { data: cashflow1 } = useCashFlow(
    account.guid,
    Interval.fromDateTimes(
      DateTime.now().minus({ month: 1 }).startOf('month'),
      DateTime.now().minus({ month: 1 }).endOf('month'),
    ),
  );
  const totalEarn0 = Math.abs(cashflow0?.filter(c => c.type === 'INCOME').reduce(
    (total, c) => c.total + total,
    0,
  ) || 0);
  const totalEarn1 = Math.abs(cashflow1?.filter(c => c.type === 'INCOME').reduce(
    (total, c) => c.total + total,
    0,
  ) || 0);
  const cashflowDifference = totalEarn0 - totalEarn1;

  return (
    <StatisticsWidget
      className="mr-2"
      title="This month income"
      stats={moneyToString(totalEarn0, currency?.mnemonic || '')}
      description={(
        <div>
          {(cashflowDifference < 0 ? '' : '+')}
          {moneyToString(cashflowDifference, currency?.mnemonic || '')}
          {' '}
          from last month
        </div>
      )}
    />
  );
}
