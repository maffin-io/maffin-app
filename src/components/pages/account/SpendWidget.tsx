import React from 'react';
import { BiCalendar } from 'react-icons/bi';
import { DateTime, Interval } from 'luxon';

import { useCashFlow } from '@/hooks/api';
import type { Account } from '@/book/entities';
import StatisticsWidget from '@/components/StatisticsWidget';
import Money from '@/book/Money';
import { isLiability } from '@/book/helpers';

export type SpendWidgetProps = {
  account: Account,
};

/**
 * Given a cashflow, compute the money that has been spent in this account
 * through EXPENSE accounts.
 */
export default function SpendWidget({
  account,
}: SpendWidgetProps): React.JSX.Element {
  const zero = new Money(0, account.commodity.mnemonic || '');

  const { data: periodCashflow } = useCashFlow(account.guid);
  const periodSpend = periodCashflow?.filter(c => c.type === 'EXPENSE').reduce(
    (total, c) => c.total.add(total),
    zero,
  ) || zero;

  const { data: currentMonthCashflow } = useCashFlow(
    account.guid,
    Interval.fromDateTimes(
      DateTime.now().startOf('month'),
      DateTime.now(),
    ),
  );
  const monthSpend = currentMonthCashflow?.filter(
    c => c.type === 'EXPENSE' || isLiability(c.type),
  ).reduce(
    (total, c) => c.total.add(total),
    zero,
  ) || zero;

  return (
    <StatisticsWidget
      className="mr-2"
      title="Expenses"
      stats={periodSpend.format()}
      description={(
        <div className="flex items-center">
          <BiCalendar className="mr-1" />
          {monthSpend.abs().format()}
          {' '}
          this month
        </div>
      )}
    />
  );
}
