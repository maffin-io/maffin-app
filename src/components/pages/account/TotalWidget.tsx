import React from 'react';

import { useAccountsTotals } from '@/hooks/api';
import Money from '@/book/Money';
import type { Account } from '@/book/entities';
import StatisticsWidget from '@/components/StatisticsWidget';
import TotalChange from '@/components/widgets/TotalChange';

export type TotalWidgetProps = {
  account: Account,
};

export default function TotalWidget({
  account,
}: TotalWidgetProps): JSX.Element {
  const { data: t0 } = useAccountsTotals();
  const total0 = t0?.[account.guid] || new Money(0, account.commodity.mnemonic);

  return (
    <StatisticsWidget
      className="mr-2"
      title="Total"
      stats={total0.format()}
      description={<TotalChange account={account} />}
    />
  );
}
