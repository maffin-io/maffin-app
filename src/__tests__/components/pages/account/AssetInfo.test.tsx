import React from 'react';
import { render } from '@testing-library/react';
import { DateTime } from 'luxon';

import { AssetInfo } from '@/components/pages/account';
import { NetWorthHistogram, AssetSankey } from '@/components/charts';
import StatisticsWidget from '@/components/StatisticsWidget';
import TotalWidget from '@/components/pages/account/TotalWidget';
import SpendWidget from '@/components/pages/account/SpendWidget';
import EarnWidget from '@/components/pages/account/EarnWidget';
import type { Account } from '@/book/entities';

jest.mock('@/components/charts/NetWorthHistogram', () => jest.fn(
  () => <div data-testid="NetWorthHistogram" />,
));

jest.mock('@/components/StatisticsWidget', () => jest.fn(
  () => <div data-testid="StatisticsWidget" />,
));

jest.mock('@/components/charts/AssetSankey', () => jest.fn(
  () => <div data-testid="AssetSankey" />,
));

jest.mock('@/components/pages/account/TotalWidget', () => jest.fn(
  () => <div data-testid="TotalWidget" />,
));

jest.mock('@/components/pages/account/SpendWidget', () => jest.fn(
  () => <div data-testid="SpendWidget" />,
));

jest.mock('@/components/pages/account/EarnWidget', () => jest.fn(
  () => <div data-testid="EarnWidget" />,
));

jest.mock('@/components/pages/accounts/AccountsTable', () => jest.fn(
  () => <div data-testid="AccountsTable" />,
));

describe('AssetInfo', () => {
  let account: Account;
  beforeEach(() => {
    jest.spyOn(DateTime, 'now').mockReturnValue(DateTime.fromISO('2023-03-01') as DateTime<true>);
    account = {
      guid: 'guid',
      name: 'Assets',
      type: 'ASSET',
      commodity: {
        mnemonic: 'EUR',
      },
    } as Account;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected when no splits', () => {
    const { container } = render(
      <AssetInfo account={account} />,
    );

    expect(NetWorthHistogram).toBeCalledWith(
      {
        height: 300,
        assetsGuid: 'guid',
        assetsLabel: 'Assets',
        hideAssets: true,
        hideLiabilities: true,
        liabilitiesGuid: '',
        showLegend: false,
      },
      {},
    );
    expect(AssetSankey).toBeCalledWith({ guid: 'guid' }, {});
    expect(TotalWidget).toBeCalledWith(
      { account },
      {},
    );
    expect(SpendWidget).toBeCalledWith(
      { account },
      {},
    );
    expect(EarnWidget).toBeCalledWith(
      { account },
      {},
    );
    expect(container).toMatchSnapshot();
  });

  it('shows account table when placeholder', () => {
    account.placeholder = true;
    render(
      <AssetInfo account={account} />,
    );

    expect(StatisticsWidget).toBeCalledTimes(1);
    expect(StatisticsWidget).toHaveBeenNthCalledWith(
      1,
      {
        className: 'mr-2',
        title: 'Subaccounts',
        description: '',
        statsTextClass: '!font-normal',
        // Don't know how to check for AccountsTable here
        stats: expect.anything(),
      },
      {},
    );
  });
});
