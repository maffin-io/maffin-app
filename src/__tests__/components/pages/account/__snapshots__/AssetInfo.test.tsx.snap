// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetInfo renders as expected when no splits 1`] = `
<div>
  <div
    class="grid grid-cols-12"
  >
    <div
      class="col-span-6"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            data-testid="TotalWidget"
          />
          <div
            data-testid="SpendWidget"
          />
          <div
            data-testid="EarnWidget"
          />
        </div>
        <div
          class="card col-span-8"
        >
          <div
            data-testid="AssetSankey"
          />
        </div>
      </div>
    </div>
    <div
      class="grid grid-cols-12 col-span-6"
    >
      <div
        class="card col-span-12"
      >
        <div
          data-testid="NetWorthHistogram"
        />
      </div>
      <div
        class="col-span-12"
      />
    </div>
  </div>
</div>
`;
